cli_ui/__init__.py

```diff
@@ -234,14 +247,56 @@ def write_and_flush(fileobj: FileObj, to_write: str) -> None:
     fileobj.flush()
 
 
+lock = RLock()
+
+
+class Buffer(io.StringIO):
+    def __init__(self, is_like_tty: bool) -> None:
+        super().__init__()
+        self.is_like_tty = is_like_tty
+
+    def flush(self) -> None:
+        # this is intentially no-op to prevent clearing the buffer before printing it
+        pass
+
+    def print(self) -> None:
+        with lock:
```

Comment:
This protect from printing 2 buffers interleaved?
Having buffers seems to allow delaying the printing to when it is wanted (or even make it conditional).
I wonder if this lock is aalway wanted. Also if its even needed does the underlying function locks? Or simply the GIL?

Reply:
Yes, it’s for making sure that outputting the buffers in parallel threads will not make the output interleaved.

Of course we could make using the lock conditional but why would we want that? Can you give an example of a use case for it?